File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = ""
        for c in s:
            if c in "aeiouAEIOU":
                vowels += c
        vowels = vowels[::-1]
        idx = 0
        res = ""
        for c in s:
            if c in "aeiouAEIOU":
                res += vowels[idx]
                idx += 1
            else:
                res += c
        return res

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.rs
impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let is_vowel = |b: u8| matches!(b, b'a' | b'e' | b'i' | b'o' | b'u' | b'A' | b'E' | b'I' | b'O' | b'U');
        let mut vb = s.into_bytes();
        let (mut i, mut j) = (0, vb.len() - 1);
        while i < j {
            while i < j && !is_vowel(vb[i]) {
                i += 1;
            }
            while i < j && !is_vowel(vb[j]) {
                j -= 1;
            }
            if i < j {
                vb.swap(i, j);
                i += 1;
                j -= 1;
            }
        }
        String::from_utf8(vb).unwrap()
    }
}